File: transektcount/src/main/java/com/wmstein/transektcount/ListSectionActivity.java
package com.wmstein.transektcount;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.WindowManager;
import android.widget.LinearLayout;
import android.widget.ListView;

import com.wmstein.transektcount.database.Section;
import com.wmstein.transektcount.database.SectionDataSource;

import java.util.List;

import androidx.appcompat.app.AppCompatActivity;

/************************************************************************************
 * Shows the list of selectable sections which is put together by SectionListAdapter.
 * Based on ListProjectActivity.java by milo on 05/05/2014.
 * Starts CountingActivity, EditSectionActivity and NewSectionActivity.
 * Changes and additions for TransektCount by wmstein since 2016-02-16,
 * last edited on 2023-07-05
 */
public class ListSectionActivity extends AppCompatActivity implements SharedPreferences.OnSharedPreferenceChangeListener
{
    //private static final String TAG = "TransektCountListSectAct";
    @SuppressLint("StaticFieldLeak")
    private static TransektCountApplication transektCount;

    // preferences
    private SharedPreferences prefs;
    private boolean brightPref;

    private SectionDataSource sectionDataSource;
    List<Section> sections;
    int maxId;
    ListView list;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_list_section);

        transektCount = (TransektCountApplication) getApplication();
        prefs = TransektCountApplication.getPrefs();
        prefs.registerOnSharedPreferenceChangeListener(this);
        brightPref = prefs.getBoolean("pref_bright", true);

        // Set full brightness of screen
        if (brightPref)
        {
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);
            WindowManager.LayoutParams params = getWindow().getAttributes();
            params.screenBrightness = 1.0f;
            getWindow().setAttributes(params);
        }

        LinearLayout list_view = findViewById(R.id.list_view);
        list_view.setBackground(transektCount.getBackground());
        list = findViewById(android.R.id.list);
    }

    public void deleteSection(Section sct)
    {
        sectionDataSource.deleteSection(sct);
        showData();
        list.invalidate(); //force list to draw
    }

    @Override
    protected void onResume()
    {
        super.onResume();

        prefs = TransektCountApplication.getPrefs();
        prefs.registerOnSharedPreferenceChangeListener(this);
        brightPref = prefs.getBoolean("pref_bright", true);

        sectionDataSource = new SectionDataSource(this);
        sectionDataSource.open();
        showData();
    }

    @Override
    protected void onPause()
    {
        super.onPause();
        sectionDataSource.close();
    }

    // show sections list
    public void showData()
    {
        Runtime.getRuntime().gc(); // garbage collection to free memory
        sections = sectionDataSource.getAllSections(prefs);
        maxId = sectionDataSource.getMaxId();
        SectionListAdapter adapter = new SectionListAdapter(this, R.layout.listview_section_row, sections, maxId);
        ListView lv = findViewById(android.R.id.list);
        lv.setAdapter(adapter);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.list_section, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        if (id == R.id.newSect)
        {
            startActivity(new Intent(this, NewSectionActivity.class).addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP));
            return true;
        }
        return super.onOptionsItemSelected(item);
    }


    public void onSharedPreferenceChanged(SharedPreferences prefs, String key)
    {
        LinearLayout list_view = findViewById(R.id.list_view);
        list_view.setBackground(null);
        list_view.setBackground(transektCount.setBackground());
        
        brightPref = prefs.getBoolean("pref_bright", true);
    }
}

File: transektcount/src/main/java/com/wmstein/transektcount/ListSectionActivity.kt
package com.wmstein.transektcount

import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.view.WindowManager
import android.widget.LinearLayout
import android.widget.ListView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NavUtils
import com.wmstein.transektcount.database.Section
import com.wmstein.transektcount.database.SectionDataSource

/************************************************************************************
 * Shows the list of selectable sections which is put together by ListSectionAdapter.
 * Based on ListProjectActivity.java by milo on 05/05/2014.
 * Starts EditSpeciesListActivity.
 * Changes and additions for TransektCount by wmstein since 2016-02-16,
 * last edited in Java on 2023-07-07,
 * converted to Kotlin on 2023-07-17,
 * last edited on 2024-08-23
 */
class ListSectionActivity : AppCompatActivity() {
    private var transektCount: TransektCountApplication? = null

    // preferences
    private var prefs = TransektCountApplication.getPrefs()
    private var brightPref = false

    private var sectionDataSource: SectionDataSource? = null
    var sections: List<Section>? = null // list of all sections
    private var maxId = 0
    var list: ListView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_list_section)
        transektCount = application as TransektCountApplication

        val listView = findViewById<LinearLayout>(R.id.list_view)
        listView.background = transektCount!!.background
        list = findViewById(android.R.id.list)

        sectionDataSource = TransektCountApplication.getSectionDS()

        // new onBackPressed logic
        if (Build.VERSION.SDK_INT >= 33) {
            onBackPressedDispatcher.addCallback(object :
                OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    NavUtils.navigateUpFromSameTask(this@ListSectionActivity)
                }
            })
        }
    }
    // end of onCreate

    override fun onResume() {
        super.onResume()

        brightPref = prefs.getBoolean("pref_bright", true)

        // Set full brightness of screen
        if (brightPref) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
            val params = window.attributes
            params.screenBrightness = 1.0f
            window.attributes = params
        }

        sectionDataSource!!.open()
        sections = sectionDataSource!!.getAllSections(prefs!!)
        maxId = sectionDataSource!!.maxId

        showData()
    }
    // end of onResume

    override fun onPause() {
        super.onPause()

        sectionDataSource!!.close()
    }

    @Deprecated("Deprecated in Java")
    @SuppressLint("ApplySharedPref", "MissingSuperCall")
    override fun onBackPressed() {
        // close the data sources
        sectionDataSource!!.close()

        NavUtils.navigateUpFromSameTask(this)
        @Suppress("DEPRECATION")
        super.onBackPressed()
    }

    // delete section as well as respective alerts
    fun deleteSection(sct: Section?) {
        sectionDataSource!!.deleteSection(sct!!)
        sections = sectionDataSource!!.getAllSections(prefs!!)
        maxId = sectionDataSource!!.maxId

        showData()
        list!!.invalidate() //force list to draw
    }

    // show sections list by ListSectionAdapter
    private fun showData() {
        // garbage collection to free memory for longer section list
//        Runtime.getRuntime().gc()
        val adapter = ListSectionAdapter(
            this,
            R.layout.listview_section_row,
            sections!!,
            maxId
            )
        val lv = findViewById<ListView>(android.R.id.list)
        lv.adapter = adapter
    }

}
